include/caf/events/module_state_event.h

```diff
@@ -60,6 +60,25 @@ static inline const void *module_id_get(size_t idx)
 	return *((&MODULE_ID_LIST_START) + idx);
 }
 
+/**
+ * @brief Get IDX of module with given id.
+ *
+ * @param[in] id ID of the module.
+ *
+ * @return Module IDX.
+ */
+static inline size_t module_idx_get(const void *id)
+{
+	for (size_t idx = 0; idx < module_count(); idx++) {
```

Comment:
```
__ASSERT_NO_MSG((id < &MODULE_ID_LIST_STOP) && (id >= &MODULE_ID_LIST_START));
if ((id >= &MODULE_ID_LIST_STOP) || (id < &MODULE_ID_LIST_START)) {
    return 0;
}
return id - &MODULE_ID_LIST_START;

```

Reply:
it doesnt work.

Reply:
Im wondering if we could possibly extent `module_state_event` to have `module_idx` as well as `module_id`

Reply:
Discussed offline. The solution suggested by @rakons would not work. We cannot get address of the id in section, so we cannot simply subtract pointers.

Reply:
The main drawback of this function is that we iterate over all of the application modules (o(n)). Pointer subtraction would be o(1), but we cannot use it here.

We could consider either:

- switching to module IDX in module_state_event (I guess it's a no go because too much existing code uses it already; it would be clean solution since we would identify modules using mainly indexes. We could also use module flags in similar situations easily)
- including module IDX next to module ID in module_state_event (it's additional 4 bytes of memory per event and it introduces duplicated information related to module identification; the good thing is that users could easily rely on module IDX to use module flags in application modules)
- Keep using module IDs in settings_loader (there are limitations related to defining array of module ids statically using MODULE_ID macro, we need to create a user-defined function to wrap the macro calls)

@pdunaj what's your opinion on this ?

Reply:
I am quite sure we can get the pointer to the variable in the section. In a matter of fact - we are doing it all the time. There may be an issue that we are using pointers to void, but still, nothing that we cannot handle.

Reply:
Ok - I can see that we are not interested in the pointer itself but the value of the pointer. This makes it little harder.